tests/test_extensions.py

```diff
@@ -0,0 +1,84 @@
+import pytest
+
+
+@pytest.fixture
+def temp_extension_registry():
+    from wsimod.extensions import extensions_registry
+
+    bkp = extensions_registry.copy()
+    extensions_registry.clear()
+    yield
+    extensions_registry.clear()
+    extensions_registry.update(bkp)
+
+
+def test_register_node_patch(temp_extension_registry):
+    from wsimod.extensions import extensions_registry, register_node_patch
+
+    # Define a dummy function to patch a node method
+    @register_node_patch("node_name.method_name")
+    def dummy_patch():
+        print("Patched method")
+
+    # Check if the patch is registered correctly
+    assert extensions_registry[("node_name.method_name", None, False)] == dummy_patch
+
+    # Another function with other arguments
+    @register_node_patch("node_name.method_name", item="default", is_attr=True)
+    def another_dummy_patch():
+        print("Another patched method")
+
+    # Check if this other patch is registered correctly
+    assert (
+        extensions_registry[("node_name.method_name", "default", True)]
+        == another_dummy_patch
+    )
+
+
+def test_apply_patches(temp_extension_registry):
+    from wsimod.arcs.arcs import Arc
+    from wsimod.extensions import (
+        apply_patches,
+        extensions_registry,
+        register_node_patch,
+    )
+    from wsimod.nodes import Node
+    from wsimod.orchestration.model import Model
+
+    # Create a dummy model
+    node = Node("dummy_node")
+    node.dummy_arc = Arc("dummy_arc", in_port=node, out_port=node)
+    model = Model()
+    model.nodes[node.name] = node
+
+    # 1. Patch a method
+    @register_node_patch("dummy_node.apply_overrides")
+    def dummy_patch():
+        pass
+
+    # 2. Patch an attribute
+    @register_node_patch("dummy_node.t", is_attr=True)
+    def another_dummy_patch(node):
+        return f"A pathced attribute for {node.name}"
+
+    # 3. Patch a method with an item
+    @register_node_patch("dummy_node.pull_set_handler", item="default")
+    def yet_another_dummy_patch():
+        pass
+
+    # 4. Path a method of an attribute
+    @register_node_patch("dummy_node.dummy_arc.arc_mass_balance")
+    def arc_dummy_patch():
+        pass
+
+    # Check if all patches are registered
+    assert len(extensions_registry) == 4
+
+    # Apply the patches
+    apply_patches(model)
+
+    # Verify that the patches are applied correctly
+    assert model.nodes[node.name].apply_overrides == dummy_patch
+    assert model.nodes[node.name].t == another_dummy_patch(node)
+    assert model.nodes[node.name].pull_set_handler["default"] == yet_another_dummy_patch
+    assert model.nodes[node.name].dummy_arc.arc_mass_balance == arc_dummy_patch
```

Comment:
Maybe we can include an example that behaves like a conventional decorator (since this will be a common use case)?

```suggestion
    # 5. Patch a decorator
    @register_node_patch("dummy_node.pull_distributed")
    def a_dummy_decorator(node, vqip):
        #Only pull from Reservoir
        return node.pull_distributed(vqip, of_type=['Reservoir'])
        
    # Check if all patches are registered
    assert len(extensions_registry) == 5

    # Apply the patches
    apply_patches(model)

    # Verify that the patches are applied correctly
    assert model.nodes[node.name].apply_overrides == dummy_patch
    assert model.nodes[node.name].t == another_dummy_patch(node)
    assert model.nodes[node.name].pull_set_handler["default"] == yet_another_dummy_patch
    assert model.nodes[node.name].dummy_arc.arc_mass_balance == arc_dummy_patch
    assert model.nodes[node.name].dummy_arc.pull_distributed == a_dummy_decorator
```

Reply:
Actually I think I've misunderstood the decorator example.

If I add the line:
```python
_ = model.nodes[node.name].pull_distributed(node.empty_vqip())
```

To the test (which would be the normal use case - it fails. doesn't seem to help if I set `is_attr=True`... so definitely an example to cover that would be helpful ;)

Reply:
I don't understand what you mean by "a conventional decorator". And where are you putting that line?

Reply:
sorry I mean - to extend the behaviour of an existing function that has some arguments.

The main new test to pass would be this:

```python
    # 5. Patch a decorator
    @register_node_patch("dummy_node.pull_distributed")
    def a_dummy_decorator(node, vqip):
        #Only pull from Reservoir
        return node.pull_distributed(vqip, of_type=['Reservoir'])
    #...
    #Test function assigned
    assert model.nodes[node.name].pull_distributed == a_dummy_decorator
    #Test new function is callable
    _ = model.nodes[node.name].pull_distributed(node.empty_vqip())
```

Reply:
OK i had a test and I realise it's not an issue with the `apply_patches` its just me not calling it properly. Thus, I just propose to add a test of extending an existing function. The below passes:

```python
    @register_node_patch("dummy_node.pull_distributed")
    def extend_function(node):
        #Only pull from Reservoir
        def f(vqip):
            return node.pull_distributed(vqip, of_type=['Reservoir'])
        return f
    
    # Apply the patches
    apply_patches(model)


    _ = model.nodes[node.name].pull_distributed(node.empty_vqip())
```

Reply:
OK sorry for all the messages there - I was just trying to figure out how to use this properly.

The below test passes, but is it the correct way to extend an existing function (while still calling it)? If so it should be in tests and in the documentation as it will be one of the more common uses of `extensions`.

```python
def assertDictAlmostEqual(d1, d2, accuracy=19):
    """

    Args:
        d1:
        d2:
        accuracy:
    """
    for d in [d1, d2]:
        for key, item in d.items():
            d[key] = round(item, accuracy)
    assert d1 == d2

def test_apply_dec(temp_extension_registry):
    from wsimod.arcs.arcs import Arc
    from wsimod.extensions import (
        apply_patches,
        extensions_registry,
        register_node_patch,
    )
    from wsimod.nodes.storage import Reservoir
    from wsimod.orchestration.model import Model

    # Create a dummy model
    node = Reservoir(name="dummy_node", initial_storage=10,capacity = 10)
    node.dummy_arc = Arc("dummy_arc", in_port=node, out_port=node)

    vq = node.pull_distributed({'volume' : 10})
    assertDictAlmostEqual(vq, node.v_change_vqip(node.empty_vqip(),10))

    model = Model()
    model.nodes[node.name] = node

    # 5. Patch a decorator
    @register_node_patch("dummy_node.pull_distributed", is_attr=True)
    def extend_function(node):
        def wrapper(node, f_old):
            def f(vqip, *args, **kw):
                return f_old(vqip, of_type = ['Node'], *args, **kw)
            return f
        #Only pull from Reservoir
        return wrapper(node, node.pull_distributed)
        
    
    # Apply the patches
    apply_patches(model)

    # Check appropriate result
    vq = model.nodes[node.name].pull_distributed({'volume' : 10})
    assertDictAlmostEqual(vq, node.empty_vqip())
```

Reply:
OK, now I get what you want. So a common use case is to use the old function you are overriding in the override itself, something like calling `super().some_function` in a child class, right?

Ok, let's see if I can figure out the most elegant way of doing it, so the user doesn't need to deal with functions, within functions, within functions...

Reply:
yes - while operational my approach is not the most elegant ;)

Reply:
Have a look now. It might be useful to re-read the docstrings to make sure the explanations are clear.